README.md
# SignalScroll
JavaFX application that communicates with alarm monitoring automation platform API to output a live view of all alarm signals currently being received. Uses Java, JavaFX, XML, JSON and CSS.
<br><br>
Login Screen:<br>
![login](https://user-images.githubusercontent.com/68821944/147366414-f3517f1c-6cd3-4b96-a476-41fadaab70c3.png)
<br><br>
Upon successful login, you have the option to choose a specific alarm receiver task and/or filter the live events.  You can leave these blank and hit the Start button to start the live event viewer:
![start](https://user-images.githubusercontent.com/68821944/147366815-9fcc93c4-f319-43fa-a852-df98ea9a32ee.png)
<br><br>
The application communicates with the automation alarm receiver platform's API every 3 seconds to check for new activity.<br>
If new activity is received, the application outputs to the live signal scroll putting most recent activity at the bottom.<br>
Higher priority alarm events are colorized so they stand out to the user.<br>
Once the events reach 500, the oldest items are purged to make room for newer events.<br>
![scrolling](https://user-images.githubusercontent.com/68821944/147366816-e0aa8ccb-43ac-4c96-87da-13c5e96e3b84.png)
<br>
<i>Image contains customer's private information so the contents have been blurred.</i>


changelog.txt
Version 1.0 - 2021-05-10
* Implementation

Version 1.1 - 2021-05-12
* Added error logging to track application process and errors
* Added code to auto update scroll to last row on refresh

Version 1.2 - 2021-05-14
* Added code to handle instances of NULL event code on signals to prevent app crash

Version 1.3 - 2021-06-10
* Adjusted code to check against "ErrorTypeNum" instead of "UserErrorMessage" for invalid session. Valid pSignalLog ErrorTypeNum results should only be 123 (0 rows returned) or 124 (x rows returned). Now if anything other than 123 or 124 is received when calling signal log, it will kick back out to login screen. Adjusted to signal scroll locking up and not displaying new results during active server swap.  It was possible the error result was something other than "Invalid Session" such as "Incorrect Password" so the code kept trying to call signal log instead of kicking out.

Version 2.0 - date 2021-06-22
* Included functionality to run signal log on just a certain task using drop down menu
* Included functionality to run signal log on just a certain search parameter from the raw data string i.e. receiver DNIS or phone number
* Added Stop button to stop the signal scroll if needed
* Added Clear button to clear the contents of the signal scroll 
* Built in functionality to accomodate Quick Response by adding line of code that sets a Central Station variable as either "AMS" or "QR". 
 - If AMS, it does normal AMS processing. If QR, it does additional processing:
	- Populates different IP address of IIS servers for server chooser
	- Populates different task drop down list for signal log search
	- Uses different color stylings of signal types
* Added drop down box to choose IIS server.  The hardcoded Central Station variable determines output of servers for drop down menu
* Updated code to accept passed server in ConnectionObject as a variable instead of using hardcoded addresses

src/signalscroll_app/Signal.java
package signalscroll_app;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import org.json.JSONObject;

/**
 *
 * @author Trey
 */
public class Signal {
    
    DateFormat df = new SimpleDateFormat("MM/dd/yy HH:mm:ss");
    
    private int seqNumBig = 0;
    private int taskNum = 0;
    private String taskDesc = "";
    private String convertedDate = "";
    private String xmit = "";
    private String eventCode = "";
    private String eventDesc = "";
    private String eventNum = "";
    private String phone = "";
    private String signalCode = ""; 
    private String point = "";
    private String rawMessage = "";
    private String siteName = "";
    private String CENTRAL_STATION = "";
    
    public Signal(JSONObject result, String cs){

        this.CENTRAL_STATION = cs;
        this.seqNumBig = result.getInt("SeqNumBig");
        this.taskNum = result.getInt("TaskNum");
        
        if (CENTRAL_STATION.compareTo("AMS") == 0) {
            this.taskDesc = findTaskDesc(result.getInt("TaskNum"));
        }
        else {
            this.taskDesc = findTaskDescQR(result.getInt("TaskNum"));
        }
        
        this.xmit = checkIfNull(result, "TransmitterCode");
        this.eventCode = checkIfNull(result, "EventCode");
        this.eventDesc = justEventDesc(checkIfNull(result, "EventCode"));
        this.eventNum = justEventNum(checkIfNull(result, "EventCode"));
        this.phone = checkIfNull(result, "Phone");
        this.signalCode = checkIfNull(result, "SignalCode");
        this.point = checkIfNull(result, "Point");
        this.siteName = checkIfNull(result, "SiteName");
        this.rawMessage = checkIfNull(result, "RawMessage");
        
        String UTCDate = result.getString("UTCDate");
        String receiveDateMilli = UTCDate.substring(UTCDate.indexOf("(")+1, UTCDate.lastIndexOf(")"));
        long milli = Long.parseLong(receiveDateMilli);
        Date date = new Date(milli);
        
        this.convertedDate = df.format(date);
        
    }

    public String getCentralStation() {
        return CENTRAL_STATION;
    }

    public String getEventCode() {
        return eventCode;
    }
    
    public String getEventDesc() {
        return eventDesc;
    }
    
    public String getEventNum() {
        return eventNum;
    }

    public String getPhone() {
        return phone;
    }

    public String getPoint() {
        return point;
    }

    public String getRawMessage() {
        return rawMessage;
    }

    public int getSeqNumBig() {
        return seqNumBig;
    }

    public String getSignalCode() {
        return signalCode;
    }

    public String getSiteName() {
        return siteName;
    }

    public int getTaskNum() {
        return taskNum;
    }

    public String getTaskDesc() {
        return taskDesc;
    }

    public String getXmit() {
        return xmit;
    }
    
    public String getConvertedDate() {
        return convertedDate;
    }
    
    private String checkIfNull (JSONObject json, String key) {
        return json.isNull(key) ? " " : json.optString(key);
    }
    
    private String justEventDesc(String event) {
        String output = "";
        try { output = event.substring(event.indexOf("-")+1);} catch (Exception ex) {}
        return output;   
    } 
    
    private String justEventNum(String event) {
        String output = "";
        try { output = event.substring(0, event.indexOf("-"));} catch (Exception ex) {}
        return output;
    } 
    
    //task description for AMS
    private String findTaskDesc(Integer taskNum){
        String output = "";
        switch (taskNum) {
            case 22:  
                output = "Twilio";
                break;
            case 300: case 301: case 302: case 303:
                output = "SG SysIII";
                break;
            case 306: case 307: case 308: case 309: 
                output = "SG SysIV";
                break;
            case 312:  
                output = "AES 1";
                break;
            case 313:  
                output = "AES 2";
                break;
            case 314:  
                output = "DMP 1";
                break;
            case 315:  
                output = "DMP 2";
                break;
            case 318:  
                output = "NAPCO 1";
                break;
            case 319:  
                output = "NAPCO 2";
                break;
            case 320:  
                output = "Uplink 1";
                break;
            case 321:  
                output = "Uplink 2";
                break;
            case 322:  
                output = "AlarmNet 1";
                break;
            case 323:  
                output = "AlarmNet 2";
                break;
            case 324:  
                output = "Alarm.Com";
                break;
            case 326:  
                output = "SecureNet";
                break;
            case 328:  
                output = "SecuraTrac";
                break;
            case 329:  
                output = "Alula";
                break;
            case 330:  
                output = "M2M 1";
                break;
            case 331:  
                output = "M2M 2";
                break;
            case 333:  
                output = "Email";
                break;
            default:
                output = "Task " + taskNum.toString();
        }
        
        return output;
    }
    
    //Task Descriptions for QuickResponse
    private String findTaskDescQR(Integer taskNum){
        String output = "";
        switch (taskNum) {
            case 18:  
                output = "Telephony";
                break;
            case 21:  
                output = "Twilio";
                break;
            case 28:  
                output = "ASAP";
                break;
            case 197:  
                output = "CheckVideo";
                break;
            case 198:  
                output = "Alarm.Com";
                break;
            case 200:  
                output = "XML";
                break;
            case 201: 
                output = "SysIII Pri";
                break;
            case 202:
                output = "SysIII Sec";
                break;
            case 203: 
                output = "D6600 Pri";
                break;
            case 204: 
                output = "D6600 Sec";
                break;
            case 205:  
                output = "AlarmNet-I";
                break;
            case 206:  
                output = "AlarmNet-XB";
                break;
            case 208:  
                output = "ITI 1";
                break;
            case 209:  
                output = "NetLink";
                break;
            case 210:  
                output = "OSB 2000 1";
                break;
            case 211:  
                output = "OH2000 2";
                break;
            case 212:  
                output = "OSB2000 3";
                break;
            case 214:  
                output = "Videofied Pri";
                break;
            case 216:  
                output = "Interlogix Pri";
                break;
            case 217:  
                output = "Interlogix Sec";
                break;
            case 218:  
                output = "Visor 1";
                break;
            case 219:  
                output = "Visor 2";
                break;
            case 220:  
                output = "SysV Pri";
                break;
            case 221:  
                output = "SysV Sec";
                break;
            case 222:  
                output = "SysII Pri";
                break;
            case 223:  
                output = "SysII Sec";
                break;
            case 224:  
                output = "DMP Pri";
                break;
            case 225:  
                output = "DMP Sec";
                break;
            case 226:  
                output = "M2M Pri";
                break;
            case 227:  
                output = "M2M Sec";
                break;
            case 228:  
                output = "AES 1 Pri";
                break;
            case 229:  
                output = "AES 1 ";
                break;
            case 230:  
                output = "Bosch Cloud";
                break;
            case 231:  
                output = "SysV Video Pri";
                break;
            case 233:  
                output = "SysIII Serial Pri";
                break;
            case 234:  
                output = "SysIII Serial Sec";
                break;
            case 235:  
                output = "SecureNet Pri";
                break;
            case 236:  
                output = "Uplink Pri";
                break;
            case 237:  
                output = "Uplink Sec";
                break;
            case 238:  
                output = "M2M TCP-IP Pri";
                break;
            case 239:  
                output = "M2M TCP-IP Sec";
                break;
            default:
                output = "Task " + taskNum.toString();
        }
        
        return output;
    }
   
}


src/signalscroll_app/InternalController.java
package signalscroll_app;

import java.io.IOException;
import java.net.URL;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.ResourceBundle;
import java.util.logging.Logger;
import javafx.application.Platform;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.fxml.Initializable;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.ComboBox;
import javafx.scene.control.Label;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableRow;
import javafx.scene.control.TableView;
import javafx.scene.control.TableView.TableViewSelectionModel;
import javafx.scene.control.TextField;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Pane;
import javafx.stage.Stage;
import org.json.JSONArray;
import org.json.JSONObject;

/**
 * FXML Controller class
 *
 * @author Trey
 */
public class InternalController implements Initializable {
    
    @FXML
    private Button logoutButton;
    
    @FXML
    private Button startButton;
    
    @FXML
    private Button stopButton;
    
    @FXML
    private Button clearButton;
    
    @FXML
    private Label taskPickerLabel;
    
    @FXML
    private Label searchLabel;
    
    @FXML
    private TextField searchTextField;
    
    @FXML
    private ComboBox taskPickerComboBox;
    
    @FXML
    private TableView signalLogTable = new TableView();
    
    @FXML
    private TableColumn colTask = new TableColumn("Task");
    
    @FXML
    private TableColumn colDate = new TableColumn("Date");
    
    @FXML
    private TableColumn colXmit = new TableColumn("Xmit");
    
    @FXML
    private TableColumn colEventCode = new TableColumn("Event Code");
    
    @FXML
    private TableColumn colSignalCode = new TableColumn("Signal Code");
    
    @FXML
    private TableColumn colPoint = new TableColumn("Point");
    
    @FXML
    private TableColumn colPhone = new TableColumn("Caller ID");
    
    @FXML
    private TableColumn colSiteName = new TableColumn("Site Name");
    
    @FXML
    private TableColumn colRawMessage = new TableColumn("Raw Message");
    
    @FXML
    private GridPane grid = new GridPane();
    
    @FXML
    private Pane pane = new Pane();
    
    @FXML
    private AnchorPane anchorPane = new AnchorPane();

    private String CENTRAL_STATION = "";
    private String IIS_SERVER = "";
    private String VALID_SN = "";
    private String VALID_PW = "";
    private int i = 0;
    private int highestSeqNum = 0;
    public ConnectionObject connObjInt = null;
    public DateFormat df = new SimpleDateFormat("MM/dd/yy HH:mm:ss");
    
    
    //quit is if manually pressing stop button
    //forceQuit is if invalid session
    Boolean quit = false;
    Boolean forceQuit = false;
    
    //create logger object
    private static Logger logger = signalscroll_app.SignalScroll_App.logger;

    
    @FXML
    public void logoutButtonHandleAction(ActionEvent event) throws IOException{
        quit = true;
        startButton.setVisible(true);   
        stopButton.setVisible(false);   
        clearButton.setVisible(false);   
        taskPickerComboBox.setDisable(false);
        searchTextField.setDisable(false);
        connObjInt.logout(VALID_SN, VALID_PW);        
        
        logger.info("Logout button pressed; Signal Scroll process stopped and user logged out.");
        
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(getClass().getResource("FXMLDocument.fxml"));
            Parent mainControllerParent = (Parent) fxmlLoader.load();
            Stage window = (Stage)((Node)event.getSource()).getScene().getWindow();
            window.setScene(new Scene(mainControllerParent));

            FXMLDocumentController mainController = fxmlLoader.<FXMLDocumentController>getController();

            window.show();
            
        } catch (Exception ex) {
            logger.severe(ex.toString());
        }
    }
    
    @FXML
    public void taskPickerComboBoxHandleAction(ActionEvent event) throws IOException{
        logger.info("Task List Changed - selected " + taskPickerComboBox.getSelectionModel().getSelectedItem());
    }
    
    @FXML
    public void invalidSessionLoadLogin() {
        quit = true;
        forceQuit = true;
        startButton.setVisible(true);       
        stopButton.setVisible(false);  
        clearButton.setVisible(false);  
        taskPickerComboBox.setDisable(false);
        searchTextField.setDisable(false);
        
        logger.info("Invalid session; signal Scroll process stopped and load login screen.");
        
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(getClass().getResource("FXMLDocument.fxml"));
            Parent mainControllerParent = (Parent) fxmlLoader.load();
            Stage window = (Stage)anchorPane.getScene().getWindow();
            window.setScene(new Scene(mainControllerParent));

            FXMLDocumentController mainController = fxmlLoader.<FXMLDocumentController>getController();

            window.show();
            
        } catch (Exception e) {
            logger.severe(e.toString());
        }
        
    }
    
    @FXML
    public void startButtonHandleAction(ActionEvent event) throws IOException{
        
        quit = false;
        forceQuit = false;
        startButton.setVisible(false);
        stopButton.setVisible(true);    
        clearButton.setVisible(false);    
        taskPickerComboBox.setDisable(true);
        searchTextField.setDisable(true);
        
        logger.info("Start Scroll button pressed.");
        try {logger.info("Passed Task Parameter: " + taskPickerComboBox.getValue().toString());} 
        catch (Exception ex){logger.info("Passed Task Parameter: All");}
        
        logger.info("Passed Search Parameter: " + searchTextField.getText());
        logger.info("Signal log started");
        signalLogTable.refresh();
        callSignalLog();
    }
    
    @FXML
    public void stopButtonHandleAction(ActionEvent event) throws IOException{
        
        quit = true;
        forceQuit = false;
        startButton.setVisible(true);
        stopButton.setVisible(false);    
        clearButton.setVisible(true);    
        taskPickerComboBox.setDisable(false);
        searchTextField.setDisable(false);
        logger.info("Stop Scroll button pressed; signal log stopped");
 
    }
    
    @FXML
    public void clearButtonHandleAction(ActionEvent event) throws IOException{
        
        signalLogTable.getItems().clear();
        signalLogTable.refresh();
        logger.info("Clear Scroll button pressed; signal log contents reset");
        
    }

    /**
     * Initializes the controller class.
     */
    @Override
    public void initialize(URL url, ResourceBundle rb) {
        
        colTask.setCellValueFactory(new PropertyValueFactory<>("taskDesc"));
        colDate.setCellValueFactory(new PropertyValueFactory<>("convertedDate"));
        colXmit.setCellValueFactory(new PropertyValueFactory<>("xmit"));
        colEventCode.setCellValueFactory(new PropertyValueFactory<>("eventDesc"));
        colSignalCode.setCellValueFactory(new PropertyValueFactory<>("signalCode"));
        colPoint.setCellValueFactory(new PropertyValueFactory<>("point"));
        colPhone.setCellValueFactory(new PropertyValueFactory<>("phone"));
        colSiteName.setCellValueFactory(new PropertyValueFactory<>("siteName"));
        colRawMessage.setCellValueFactory(new PropertyValueFactory<>("rawMessage"));
        
        clearButton.setVisible(false);
        
    }   
        
    public void loadTaskList() {
        
        logger.info("Loading task list from stages procedure call");
        
        taskPickerComboBox.getItems().add(0, "All");
        
        JSONObject getReq = connObjInt.taskList(VALID_SN, VALID_PW);
        
        //check for successful return
        JSONArray resultSet = getReq.getJSONArray("ResultSet");    
        int resultSetLen = resultSet.length();

        if (resultSetLen > 0) {
            
            //if CS is AMS do this list
            if (CENTRAL_STATION.compareTo("AMS") == 0) {
                for (int i = 0; i < resultSetLen; i++) {

                    int size = taskPickerComboBox.getItems().size();

                    JSONObject newTask = resultSet.getJSONObject(i);
                    Integer taskNum = newTask.getInt("TaskNum");
                    String taskDesc = newTask.getString("TaskDescription"); 

                    if (taskNum == 22) {
                        taskPickerComboBox.getItems().add(size, "" + taskNum + " - " + taskDesc); 

                    }
                    else if (taskNum >= 300 && taskNum <= 9200) {
                        taskPickerComboBox.getItems().add(size, "" + taskNum + " - " + taskDesc); 
                    }
                    else {}

                }
            }
            
            //if CS is QR do this list
            else {
                for (int i = 0; i < resultSetLen; i++) {

                    int size = taskPickerComboBox.getItems().size();

                    JSONObject newTask = resultSet.getJSONObject(i);
                    Integer taskNum = newTask.getInt("TaskNum");
                    String taskDesc = newTask.getString("TaskDescription"); 

                    if (taskNum == 18) {
                        taskPickerComboBox.getItems().add(size, "" + taskNum + " - " + taskDesc); 

                    }
                    else if (taskNum == 21) {
                        taskPickerComboBox.getItems().add(size, "" + taskNum + " - " + taskDesc); 

                    }
                    else if (taskNum == 28) {
                        taskPickerComboBox.getItems().add(size, "" + taskNum + " - " + taskDesc); 

                    }
                    else if (taskNum >= 190 && taskNum <= 9200) {
                        taskPickerComboBox.getItems().add(size, "" + taskNum + " - " + taskDesc); 
                    }
                    else {}

                }
            }
            
        }
        
        logger.info("Task list loaded and populated");
    
    }
    
    
    public void callSignalLog(){
        
        new Thread(() -> {
            
            while (!quit) {
                
                String search = "";
                String task = "";
                
                try {search = searchTextField.getText();} catch (Exception ex){}
                try {task = taskPickerComboBox.getValue().toString();} catch (Exception ex){}
                
                DateFormat dfHHmm = new SimpleDateFormat("HHmm");
                DateFormat dfStartDay = new SimpleDateFormat("yyyy/MM/dd 00:00:00");
                Date currentDateTime = new Date(System.currentTimeMillis());
                String outputDateStartDay = dfStartDay.format(currentDateTime);
                String passedStartHHMM = "";

                Date date2 = new Date();
                long newMilli = 0;
                long milliTimeZoneOffset = 18000000; //offset 5 hours

                try {
                    date2 = dfStartDay.parse(outputDateStartDay);
                    newMilli = date2.getTime() - milliTimeZoneOffset;
                    passedStartHHMM = dfHHmm.format(currentDateTime);
                } catch (ParseException ex) {
                    logger.severe("Parse Exception: " + ex.toString());
                } 
                
                
                /********** debug mode *********/
                //date2 = dfStartDay.parse("2021/05/14 08:00:00");
                //newMilli = 1620950400000L;
                //passedStartHHMM = "0255";
                /*******************************/
                
                JSONObject getReq = connObjInt.signalLog(VALID_SN, VALID_PW, newMilli, passedStartHHMM, search, task);
                            
                //get ErrorTypeNum value - 123 is 0 rows returned, 124 is X rows returned
                Integer errorCode = getReq.getInt("ErrorTypeNum");
                
                //checks for a valid session.. if valid, it continues
                if (errorCode == 123 || errorCode == 124) {
                    
                    //check for successful return
                    JSONArray resultSet = getReq.getJSONArray("ResultSet");    
                    int resultSetLen = resultSet.length();

                    if (resultSetLen > 0) {

                        for (int i = resultSetLen-1; i >= 0 ; i--) {


                            JSONObject newSite = resultSet.getJSONObject(i);
                            Integer seqNum = newSite.getInt("SeqNumBig");

                            if (seqNum > highestSeqNum) {

                                //if signal lob table is over 500 items, remove item to keep list small
                                if (signalLogTable.getItems().size() > 500) {
                                    signalLogTable.getItems().remove(0);
                                }

                                highestSeqNum = seqNum;

                                signalLogTable.setRowFactory(tv -> new TableRow<Signal>(){

                                    @Override
                                    protected void updateItem(Signal item, boolean empty) {

                                        try {
                                            super.updateItem(item,empty);

                                            String eventNum = item.getEventNum();
                                            
                                            //if chosen CS is AMS do this styling
                                            if (CENTRAL_STATION.compareTo("AMS") == 0) {
                                                switch (eventNum) {
                                                    case "001": case "020":
                                                        getStyleClass().clear();
                                                        getStyleClass().add("alarm-type");
                                                        getStyleClass().add("fire");
                                                        break;
                                                    case "002": case "016":  case "017": case "034": 
                                                        getStyleClass().clear();
                                                        getStyleClass().add("panic");
                                                        getStyleClass().add("alarm-type");
                                                        break;
                                                    case "003": case "019":  case "061": 
                                                        getStyleClass().clear();
                                                        getStyleClass().add("burg");
                                                        getStyleClass().add("alarm-type");
                                                        break;
                                                    case "004": 
                                                        getStyleClass().clear();
                                                        getStyleClass().add("medic");
                                                        getStyleClass().add("alarm-type");
                                                        break;

                                                    default:
                                                        getStyleClass().clear();
                                                        getStyleClass().add("cell");
                                                        getStyleClass().add("indexed-cell");
                                                        getStyleClass().add("table-row-cell");

                                                }
                                            }
                                            
                                            //if chosen CS is QR do this styling
                                            else {
                                                switch (eventNum) {
                                                    case "001": case "024":  case "030":  case "110":  case "111": 
                                                    case "112": case "113":  case "114":  case "115": 
                                                    case "116": case "117":  case "024N":  case "024Pro": 
                                                    case "FA": case "Fire":  case "GA":  
                                                        getStyleClass().clear();
                                                        getStyleClass().add("alarm-type");
                                                        getStyleClass().add("fireQR");
                                                        break;
                                                    case "002": case "016":  case "017":
                                                        getStyleClass().clear();
                                                        getStyleClass().add("panicQR");
                                                        getStyleClass().add("alarm-type");
                                                        break;
                                                    case "003": case "130":  case "131": 
                                                    case "132": case "133":  case "134": 
                                                    case "003D": case "003V":  case "BA":
                                                        getStyleClass().clear();
                                                        getStyleClass().add("burgQR");
                                                        getStyleClass().add("alarm-type");
                                                        break;
                                                    case "004": case "031": case "004D": 
                                                        getStyleClass().clear();
                                                        getStyleClass().add("medicQR");
                                                        getStyleClass().add("alarm-type");
                                                        break;
                                                    case "019": case "026": case "027": 
                                                    case "028": case "029": case "026G": 
                                                        getStyleClass().clear();
                                                        getStyleClass().add("criticalQR");
                                                        getStyleClass().add("alarm-type");
                                                        break;

                                                    default:
                                                        getStyleClass().clear();
                                                        getStyleClass().add("cell");
                                                        getStyleClass().add("indexed-cell");
                                                        getStyleClass().add("table-row-cell");

                                                }
                                            }
                                        }
                                        catch (NullPointerException ex) {
                                            //nothing to do here.. null is to be expected sometimes
                                        }

                                    }
                                });

                                signalLogTable.getItems().add(new Signal(newSite, CENTRAL_STATION));

                                int lastRow = signalLogTable.getItems().size();
                                
                                Platform.runLater(new Runnable() {
                                    @Override
                                    public void run() {
                                        TableViewSelectionModel selectionModel = signalLogTable.getSelectionModel();
                                        signalLogTable.scrollTo(lastRow);
                                        signalLogTable.scrollToColumnIndex(0);
                                        selectionModel.select(lastRow, colTask);
                                    }
                                });

                            }
                        }
                    }

                    try {
                        Thread.sleep(3000);
                    } catch (InterruptedException ex) {
                        logger.severe("Interrupted Exception while sleeping: " + ex.toString());
                    }
                    
                }
                
                //invalid session - exit the loop and thread.
                else {
                    logger.info("ErrorTypeNum:" + errorCode);
                    logger.info("UserErrorMessage: " + getReq.get("UserErrorMessage").toString());
                    logger.info("Invalid session. Exit Signal Scroll loop");
                    quit = true;
                    forceQuit = true;
                }
                
                
                
          
            }
            
            //only reach this step IF the loop ended
            Platform.runLater(new Runnable() {
                @Override public void run() {
                    
                    //if forceQuit true means there was an invalid session and load login screen
                    //otherwise it was a manual quit by user and login screen doesn't need to 
                    //load but stay internal
                    if (forceQuit) {
                        invalidSessionLoadLogin();
                    }
                    
                }
            });
            
        }).start();
        
        
        
    }
   
    public void acceptValidatedSessionInfo (String vSN, String vPW, String cs, String iis) {
        VALID_SN = vSN;
        VALID_PW = vPW;
        CENTRAL_STATION = cs;
        IIS_SERVER = iis;
        connObjInt = new ConnectionObject(cs, iis);
    }
    
    public void manualCloseWindow() {
        quit = true;
        startButton.setVisible(true);
        stopButton.setVisible(false);
        clearButton.setVisible(false);
        connObjInt.logout(VALID_SN, VALID_PW);
        logger.info("Window closed; Signal Scroll process stopped and user logged out.");
        Platform.exit();
    }
    
    private String checkIfNull (JSONObject json, String key) {
        return json.isNull(key) ? "" : json.optString(key);
    }
    
    
}
